### ISSUE TITLE
48-bit userspace VA on AArch64 breaks freelist implementation

### ISSUE THREAD

**Issue by linnaea** [apache/trafficserver#5016]
ARMv8 without Large System Extension does not have 128-bit CAS capability.

On such systems [include/tscore/ink_queue.h#L127](https://github.com/apache/trafficserver/blob/master/include/tscore/ink_queue.h#L127) uses the bottom 48 bits of data field to store a pointer, and sign extends it when extracted.

However on AArch64 userspace can have a 48-bit VA, the highest bit of the stored pointer gets misinterpreted as a sign bit, and the extracted pointer gets mangled, causing a segfault.

```
Thread 2 "test_freelist" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xffffbe9de1b0 (LWP 487)]
0x0000ffffbf68ec2c in freelist_free (f=f@entry=0xaaaaaaaec2e0,
    item=0xffffaaaaaaafc000) at ink_queue.cc:319
319         *adr_of_next = FREELIST_POINTER(h);
(gdb) print/x adr_of_next
$1 = 0xffffaaaaaaafc000
(gdb) frame 1
#1  0x0000ffffbf68eeb8 in freelist_new (f=0xaaaaaaaec2e0) at ink_queue.cc:239
239             freelist_free(f, a);
(gdb) print/x item.data
$2 = 0xaaaaaaafc000
(gdb)
```

**Comment by linnaea** [comment]
BTW there's also something that looks strage about the second half of that `FREELIST_POINTER` macro.

```c
((
  (
    ~(
      (
        ((intptr_t)(_x).data) << 16 >> 63     // Gives -1 if negative, 0 otherwise
      ) - 1                                   // Changes to -2 or -1
    )                                         // Negation results in 1 if negative, 0 otherwise
  ) >> 48                                     // Wouldn't this always yield 0?
) << 48)
```

**Comment by dmorilha** [comment]
Thanks for the report, I have a PR open: https://github.com/apache/trafficserver/pull/5444, would you be able to cherry-pick it, build and test on your side? Thank you @linnaea

**Comment by oknet** [comment]
Fixed by PR #5451 .
